### PR TITLE
JAMES-2340 Copying messages out of an empty mailbox should not throw

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

**Comment by aduprat** [comment]
```
Tests in error: 
  SmtpAuthIntegrationTest.authenticatedSmtpSessionsShouldBeDelivered:111 » ConditionTimeout
```
test this please

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java [hunk 1]
```diff
@@ -731,21 +731,24 @@ private MoveResult move(Iterator<MailboxMessage> originalRows, MailboxSession se
     private SortedMap<MessageUid, MessageMetaData> copy(MessageRange set, StoreMessageManager to, MailboxSession session) throws MailboxException {
         IteratorWrapper<MailboxMessage> originalRows = new IteratorWrapper<>(retrieveOriginalRows(set, session));
 
-        SortedMap<MessageUid, MessageMetaData> copiedUids = collectMetadata(to.copy(originalRows, session));
+        if (originalRows.hasNext()) {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: I don't agree with this if.
All the code works great when originalRows is empty but the event broadcasting, we should just prevent firing the events on Noop, WDYT ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java [hunk 2]
```diff
@@ -734,42 +733,42 @@ private MoveResult move(Iterator<MailboxMessage> originalRows, MailboxSession se
...
         }
         dispatcher.added(session, copiedUids, to.getMailboxEntity(), messagesMap.build());
-        dispatcher.moved(session,
-            MessageMoves.builder()
-                .previousMailboxIds(getMailboxEntity().getMailboxId())
-                .targetMailboxIds(to.getMailboxEntity().getMailboxId(), getMailboxEntity().getMailboxId())
-                .build(),
-            messagesMap.build());
-
+        if (!messagesMap.build().isEmpty()) {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: sorry if I have been unclear but I think that it's not the responsibility of move method to do that filtering.
We should be able to build the MessageMoves message with an empty list and the dispatcher should do the filtering.
We could, for example, add a "isNoop" on MessageMoves (and other evenets) to help filtering events.

Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: Sure, that would be less invasive.

If no-one do it before, I will do it on Thursday.

**Comment by chibenwa** [comment]
```
[df0726e00d9b266e14324d87cf6e1ae2c0fe3d7a] Tests in error: 
[df0726e00d9b266e14324d87cf6e1ae2c0fe3d7a]   SpamAssassinContract.movingAMailToTrashShouldNotImpactSpamassassinLearning:354 » ConditionTimeout
```

test this please

**Comment by mbaechler** [comment]
I pushed a fixup about the comment I did

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
```
memory
[db5ec6a106ae6b5190afccbb953a5c7f4208be3f] Tests in error: 
[db5ec6a106ae6b5190afccbb953a5c7f4208be3f]   VacationIntegrationTest.jmapVacationShouldNotSendNotificationTwice:236 » ConditionTimeout
```

test this please

**Comment by chibenwa** [comment]
Merged

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**